tests/infer/test_valid_models.py

```diff
@@ -879,3 +879,103 @@ def model():
+            assert x.shape == torch.Size((5, 1, 1))
+            with pyro.iarange("iarange_inner_1", 11, 6):
+                y = pyro.sample("y", dist.Bernoulli(p))
+                # allocated dim is rightmost available, i.e. -1
+                assert y.shape == torch.Size((5, 1, 6))
+                with pyro.iarange("iarange_inner_2", 12, 7):
+                    z = pyro.sample("z", dist.Bernoulli(p))
+                    # allocated dim is next rightmost available, i.e. -2
+                    assert z.shape == torch.Size((5, 7, 6))
+                    # since dim -3 is already allocated, use dim=-4
+                    with pyro.iarange("iarange_inner_3", 13, 8):
+                        q = pyro.sample("q", dist.Bernoulli(p))
+                        assert q.shape == torch.Size((8, 5, 7, 6))
+
+    guide = config_enumerate(model) if Elbo is TraceEnum_ELBO else model
+    assert_ok(model, guide, Elbo())
+
+
+@pytest.mark.parametrize("Elbo", [Trace_ELBO, TraceGraph_ELBO, TraceEnum_ELBO])
+def test_dim_allocation_error(Elbo):
+
+    @poutine.broadcast
+    def model():
+        p = torch.tensor(0.5, requires_grad=True)
+        with pyro.iarange("iarange_outer", 10, 5, dim=-2):
+            x = pyro.sample("x", dist.Bernoulli(p))
+            assert x.shape == torch.Size((5, 1))
+            # allocated dim is rightmost available, i.e. -1
+            with pyro.iarange("iarange_inner_1", 11, 6):
+                y = pyro.sample("y", dist.Bernoulli(p))
+                assert y.shape == torch.Size((5, 6))
+                # throws an error as dim=-1 is already occupied
+                with pyro.iarange("iarange_inner_2", 12, 7, dim=-1):
+                    pyro.sample("z", dist.Bernoulli(p))
+
+    guide = config_enumerate(model) if Elbo is TraceEnum_ELBO else model
+    assert_error(model, guide, Elbo())
+
+
+@pytest.mark.parametrize("Elbo", [Trace_ELBO, TraceGraph_ELBO, TraceEnum_ELBO])
+def test_vectorized_num_particles(Elbo):
+    data = torch.ones(1000, 2)
+
+    def model():
+        with pyro.iarange("components", 2):
+            p = pyro.sample("p", dist.Beta(torch.tensor(1.1), torch.tensor(1.1)))
+            assert p.shape == torch.Size((10, 1, 2))
+            with pyro.iarange("data", data.shape[0]):
+                pyro.sample("obs", dist.Bernoulli(p), obs=data)
+
+    def guide():
+        with pyro.iarange("components", 2):
+            pyro.sample("p", dist.Beta(torch.tensor(1.1), torch.tensor(1.1)))
+
+    pyro.clear_param_store()
+    guide = config_enumerate(guide) if Elbo is TraceEnum_ELBO else guide
+    assert_ok(model, guide, Elbo(num_particles=10,
+                                 vectorize_particles=True,
+                                 max_iarange_nesting=2,
+                                 strict_enumeration_warning=False))
+
+
+@pytest.mark.parametrize('enumerate_', [None, "sequential", "parallel"])
+def test_enum_discrete_vectorized_num_particles(enumerate_):
+    num_particles = 50
```

Comment:
Can you `@pytest.mark.parametrize('num_particles', [1, 10])` to test the single-particle edge case?

Reply:
Good idea, will do.